src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py

```diff
@@ -128,8 +128,14 @@ def create_role_assignment(cmd, role, assignee=None, assignee_object_id=None, re
                            scope=None):
     if bool(assignee) == bool(assignee_object_id):
         raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
-    return _create_role_assignment(cmd.cli_ctx, role, assignee or assignee_object_id, resource_group_name, scope,
-                                   resolve_assignee=(not assignee_object_id))
+
+    try:
+        return _create_role_assignment(cmd.cli_ctx, role, assignee or assignee_object_id, resource_group_name, scope,
+                                       resolve_assignee=(not assignee_object_id))
+    except Exception as ex:  # pylint: disable=broad-except
+        if _error_caused_by_role_assignment_exists(ex):  # for idempotent
+            return list_role_assignments(cmd, assignee, role, resource_group_name, scope)[0]
+        raise
```

Comment:
This implementation of idempotence has an issue.

If `--assignee-object-id` is used and if the role assignment is newly created, the returned role assignment is correct:

```
> az role assignment create --assignee-object-id e323d08e-9b5f-450e-8500-342136c6bfd5 --assignee-principal-type ServicePrincipal --role Reader --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590
{
  "canDelegate": null,
  "condition": null,
  "conditionVersion": null,
  "description": null,
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleAssignments/41030363-55bb-4b7f-8613-23d043620db5",
  "name": "41030363-55bb-4b7f-8613-23d043620db5",
  "principalId": "e323d08e-9b5f-450e-8500-342136c6bfd5",
  "principalType": "ServicePrincipal",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

But if the role assignment already exists (results in a 409 HTTP response), the returned role assignment could be wrong:

```
> az role assignment create --assignee-object-id e323d08e-9b5f-450e-8500-342136c6bfd5 --assignee-principal-type ServicePrincipal --role Reader --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590
{
  "canDelegate": null,
  "condition": null,
  "conditionVersion": null,
  "description": null,
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleAssignments/332cf847-f85f-4e42-93c0-ac2d2269c9c6",
  "name": "332cf847-f85f-4e42-93c0-ac2d2269c9c6",
  "principalId": "5963f50c-7c43-405c-af7e-53294de76abd",
  "principalName": [admin2@AzureSDKTeam.onmicrosoft.com](mailto:admin2@AzureSDKTeam.onmicrosoft.com),
  "principalType": "User",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "roleDefinitionName": "Reader",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590",
  "type": "Microsoft.Authorization/roleAssignments"
}

```

The reason is because `list_role_assignments` doesn’t take `assignee_object_id` into consideration, so it actually returns the first role assignment object matching only specified **role** and **scope**.

Meanwhile, `az role assignment list` doesn’t support `--assignee-object-id` either, which is also inconsistent with `az role assignment create`.